### DISCUSSION TITLE
How to reconstruct() from an IVF index use add_with_ids

### DISCUSSION THREAD

**Discussion by czpmango** [facebookresearch/faiss#3986] [General]
I got the error `DirectMap.cpp:82: direct map not initialized` when `reconstruct()` from an IVF index.

```
import faiss
quantizer = faiss.IndexFlatL2(dim)
index = faiss.IndexIVFFlat(quantizer, dim, nb)
index.train(train_vectors)

index.add_with_ids(db_vectors, idx)
index.make_direct_map(False) # use nomap type
item0 = index.reconstruct(0)
print("Reconstructed item 0: ", item0)
```
Is there any solution for this scenario?

**Comment by bshethmeta** [comment]
Can you help me understand why setting make_direct_map to True will not work for your usecase?

**Reply by czpmango** [reply]
I got this error `DirectMap.cpp:106: cannot have array direct map and add with ids`

**Comment by czpmango** [comment]
index.set_direct_map_type(faiss.DirectMap.Hashtable)  # set direct map

**Comment by bshethmeta** [comment]
I am not a 100% sure I understand your usecase but here is some test code that works:

d = 128
nb = 100  # number of database vectors
nd = 1000
idx = np.arange(nb).astype(np.int64)

train_vectors = np.random.random((nb, d)).astype('float32')
db_vectors = np.random.random((nb, d)).astype('float32')

quantizer = faiss.IndexFlatL2(d)
index = faiss.IndexIVFFlat(quantizer, d, 5)
index.make_direct_map()
index.set_direct_map_type(faiss.DirectMap.Hashtable)

index.train(train_vectors)
index.add_with_ids(db_vectors, idx)
ids_remove = np.array([3], dtype=np.int64)
index.remove_ids(ids_remove)

# # We should find this
y = index.reconstruct(10)
